anvil-server/src/main/kotlin/com/avail/anvil/io/Coding.kt

```diff
@@ -0,0 +1,548 @@
+	done: DoneWriting)
+{
+	var residue = this
+	while (bytes.hasRemaining())
+	{
+		// Write 7 LSB bits at a time, setting the MSB for nonfinal bytes.
+		val b = residue and 0x7F
+		residue = residue ushr 7
+		if (residue == 0L)
+		{
+			bytes.put(b.toByte())
+			return done(bytes)
+		}
+		bytes.put((b or 0x80).toByte())
+	}
+	// There wasn't enough room to fully encode the value.
+	writeMore(bytes) { bytes1 -> residue.vlq(bytes1, writeMore, done)}
+}
+
+/**
+ * Unapply the variable-length universal coding strategy applied by
+ * [`vlq`][Int.vlq] to decode a nonnegative integer from the specified buffer.
+ *
+ * @author Todd L Smith &lt;todd@availlang.org&gt;
+ * @param bytes
+ *   The source buffer.
+ * @param partial
+ *   A partially decoded integer. Defaults to `0`.
+ * @param shift
+ *   The shift factor so far. Defaults to `0`.
+ * @param readMore
+ *   How to keep reading if the source buffer exhausts prematurely.
+ * @param done
+ *   What to do with the fully decoded integer.
+ */
+internal fun unvlqInt (
+	bytes: ByteBuffer,
+	partial: Int = 0,
+	shift: Int = 0,
+	readMore: ReadMore,
+	done: DoneReading<Int>)
+{
+	var n = partial
+	var k = shift
+	while (bytes.hasRemaining())
+	{
+		val b = bytes.get().toInt()
+		when
+		{
+			b and 0x80 == 0x80 ->
+			{
+				n = n or ((b and 0x7F) shl k)
+				k += 7
+			}
+			else ->
+			{
+				// The MSB is clear, so we're done decoding.
+				return done(n or (b shl k), bytes)
+			}
+		}
+	}
+	// There weren't enough bytes to fully decode the value.
+	readMore { bytes1 -> unvlqInt(bytes1, n, k, readMore, done) }
+}
+
+/**
+ * Unapply the variable-length universal coding strategy applied by
+ * [`vlq`][Long.vlq] to decode a nonnegative integer from the specified buffer.
+ *
+ * @author Todd L Smith &lt;todd@availlang.org&gt;
+ * @param bytes
+ *   The source buffer.
+ * @param partial
+ *   A partially decoded integer. Defaults to `0`.
+ * @param shift
+ *   The shift factor so far. Defaults to `0`.
+ * @param readMore
+ *   How to keep reading if the source buffer exhausts prematurely.
+ * @param done
+ *   What to do with the fully decoded integer.
+ */
+internal fun unvlqLong (
+	bytes: ByteBuffer,
+	partial: Long = 0L,
+	shift: Int = 0,
+	readMore: ReadMore,
+	done: DoneReading<Long>)
+{
+	var n = partial
+	var k = shift
+	while (bytes.hasRemaining())
+	{
+		val b = bytes.get().toInt()
+		when
+		{
+			b and 0x80 == 0x80 ->
+			{
+				n = n or ((b and 0x7F).toLong() shl k)
+				k += 7
+			}
+			else ->
+			{
+				// The MSB is clear, so we're done decoding.
+				return done(n or (b shl k).toLong(), bytes)
+			}
+		}
+	}
+	// There weren't enough bytes to fully decode the value.
+	readMore { bytes1 -> unvlqLong(bytes1, n, k, readMore, done) }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//                              Integer coding.                               //
+////////////////////////////////////////////////////////////////////////////////
```

Comment:
This entire section is based upon a special encoding strategy, _zigzag_. Perhaps adding a _see_ link to [https://en.wikipedia.org/wiki/Variable-length_quantity#Zigzag_encoding](https://en.wikipedia.org/wiki/Variable-length_quantity#Zigzag_encoding) might be useful to viewers of the code.

Reply:
I added `@see` references for both VLQ and zigzag in all appropriate implementations.